Objects/descrobject.c

```diff
@@ -1781,38 +1781,43 @@ property_init_impl(propertyobject *self, PyObject *fget, PyObject *fset,
     Py_XINCREF(fget);
     Py_XINCREF(fset);
     Py_XINCREF(fdel);
-    Py_XINCREF(doc);
 
     Py_XSETREF(self->prop_get, fget);
     Py_XSETREF(self->prop_set, fset);
     Py_XSETREF(self->prop_del, fdel);
-    Py_XSETREF(self->prop_doc, doc);
+    Py_XSETREF(self->prop_doc, NULL);
     Py_XSETREF(self->prop_name, NULL);
 
     self->getter_doc = 0;
+    PyObject *prop_doc = NULL;
 
+    if (doc != NULL && doc != Py_None) {
+        prop_doc = doc;
+        Py_XINCREF(prop_doc);
+    }
     /* if no docstring given and the getter has one, use that one */
-    if ((doc == NULL || doc == Py_None) && fget != NULL) {
-        PyObject *get_doc;
-        int rc = _PyObject_LookupAttr(fget, &_Py_ID(__doc__), &get_doc);
+    else if (fget != NULL) {
+        int rc = _PyObject_LookupAttr(fget, &_Py_ID(__doc__), &prop_doc);
         if (rc <= 0) {
             return rc;
         }
+        self->getter_doc = 1;
```

Comment:
Looks like this attribute affects copying properties (it's used in property_copy). Could you test whether that works as expected? In particular it's odd that we set this even if prop_doc is Py_None here.

Reply:
Ok. I think I partially understand the purpose of `getter_doc` attribute. It allows overriding docstring of property copy with docstring of new getter unless the original docstring was provided explicitly at the moment of construction of the original property.

I can add the following snippet to the test set. @JelleZijlstra Does it looks right to you?

```python
    @unittest.skipIf(sys.flags.optimize >= 2,
                     "Docstrings are omitted with -O2 and above")
    def test_docstring_copy2(self):
        def getter1(self):
            return 1
        def getter2(self):
            """doc 2"""
            return 2
        def getter3(self):
            """doc 3"""
            return 3

        # Case-1: user-provied doc is preserved in copies
        p = property(getter1, None, None, "doc-A")

        p2 = p.getter(getter2)
        self.assertEqual(p.__doc__, "doc-A")
        self.assertEqual(p2.__doc__, "doc-A")

        # Case-2: with no user-provied doc new getter doc
        #         takes precendence
        p = property(getter2, None, None, None)

        p2 = p.getter(getter3)
        self.assertEqual(p.__doc__, "doc 2")
        self.assertEqual(p2.__doc__, "doc 3")

        # Case-3: user-provied doc after construction
        #         doesn't affect __doc__ overriding rule
        p = property(getter2, None, None, None)
        p.__doc__ = "my doc"
        p2 = p.getter(getter3)
        self.assertEqual(p.__doc__, "my doc")
        self.assertEqual(p2.__doc__, "doc 3")
```